zhaquirks/danfoss/thermostat.py

```diff
@@ -26,47 +26,171 @@
+    0x404B: ("regulation_setpoint_offset", t.int8s, "rpw"),
+    0x404C: ("adaptation_run_control", t.enum8, "rw"),
+    0x404D: ("adaptation_run_status", t.bitmap8, "rp"),
+    0x404E: ("adaptation_run_settings", t.bitmap8, "rw"),
+    0x404F: ("preheat_status", t.Bool, "rp"),
+    0x4050: ("preheat_time", t.uint32_t, "rp"),
+}
+# ZCL Attributes Supported: pi_heating_demand (0x0008)
+
+# reading mandatory ZCL attribute 0xFFFD results in UNSUPPORTED_ATTRIBUTE
+# ZCL Commands Supported: SetWeeklySchedule (0x01), GetWeeklySchedule (0x02), ClearWeeklySchedule (0x03)
+
+# Danfos says they support the following, but Popp eT093WRO responds with UNSUPPORTED_ATTRIBUTE
+#    0x0003, 0x0004, 0x0015, 0x0016, 0x0025, 0x0030, 0x0020, 0x0021, 0x0022
+
+# 0x0204
+danfoss_interface_attr = {
+    0x4000: ("viewing_direction", t.enum8, "rpw"),
+}
+
+# Writing to mandatory ZCL attribute 0x0000 doesn't seem to do anything
+# ZCL Attributes Supported: keypad_lockout (0x0001)
+
+# 0x0b05
+danfoss_diagnostic_attr = {
+    0x4000: ("sw_error_code", t.bitmap16, "rp"),
+    0x4001: ("wake_time_avg", t.uint32_t, "rp"),  # always 0?
+    0x4002: ("wake_time_max_duration", t.uint32_t, "rp"),  # always 0?
+    0x4003: ("wake_time_min_duration", t.uint32_t, "rp"),  # always 0?
+    0x4004: ("sleep_postponed_count_avg", t.uint32_t, "rp"),  # always 0?
+    0x4005: ("sleep_postponed_count_max", t.uint32_t, "rp"),  # always 0?
+    0x4006: ("sleep_postponed_count_min", t.uint32_t, "rp"),  # always 0?
+    0x4010: ("motor_step_counter", t.uint32_t, "rp"),
+}
+
+
+async def read_attributes(dest, source, dictionary):
+    """Automatically reads attributes from source cluster and stores them in the dest cluster."""
+    response = {}
+    step = 14  # The device doesn't respond to more than 14 per request it seems
+
+    # read from source
+    for a in range(0, len(dictionary) + step, step):
+        subset = list(dictionary.keys())[a : a + step]
+        if subset:
+            response.update(
+                (await source.read_attributes(subset, manufacturer=MANUFACTURER))[0]
+            )
+
+    # store all of them in dest
+    for attrid, value in response.items():
+        dest.update_attribute(attrid, value)
+
+
+class DanfossTRVCluster(CustomCluster, ManufacturerSpecificCluster):
+    """Danfoss custom TRV cluster."""
+
+    cluster_id = 0xFC03
+    ep_attribute = "danfoss_trv_cluster"
+
+    attributes = ManufacturerSpecificCluster.attributes.copy()
+    attributes.update(danfoss_thermostat_attr)
+
+    async def write_attributes(self, attributes, manufacturer=None):
+        """Write attributes to thermostat cluster."""
+        return await self.endpoint.thermostat.write_attributes(attributes, manufacturer)
+
+    async def bind(self):
+        """Read attributes before ZHA binds, this makes sure the entity is created."""
+        await read_attributes(self, self.endpoint.thermostat, danfoss_thermostat_attr)
```

Comment:
Hmm. I think it would be better if we have the attributes that need to be read when pairing in ``ZCL_INIT_ATTRS`` here: https://github.com/home-assistant/core/blob/dev/homeassistant/components/zha/core/channels/manufacturerspecific.py

Reply:
Fixed now. Thanks a lot